ansible/docker.yml

```diff
@@ -48,5 +48,5 @@
       changed_when: "'active' not in swarm_status.stdout_lines"
     - name: join worker to swarm
       shell: >
-        docker swarm join --token={{ swarm_token }} {{ manager_ip }}:2377
+        docker swarm leave || true && docker swarm join --token={{ swarm_token_stdout }} {{ manager_ip }}:2377
```

Comment:
Why the `|| true &&` instead of `;` ? Will semi-colon not work in ansible?
If it doesn't, can you put the `docker swarm leave` on one line, and then the join on the next line?
It just seems like there should be a better way to express executing the first command and not caring about its return code, and then executing a second command.

Reply:
`docker swarm join` doesn't have a `--force` flag, and errors out if the node is already part of a swarm (so subsequent runs of the ansible state fail since this isn't idempotent). To make this idempotent, we issue a `leave` followed by a `join`. However `leave` fails with a non-zero exit code when the node isn't already part of a swarm failing the execution of the entire shell section. The `|| true` here basically tells the shell to ignore the exit code of `docker swarm leave` and continue executing the rest. `;` wouldn't work here since it just joins the two commands together, and the shell still errors when the first command before the `;` exits with a non-zero exit code. The other alternative here would be to explicitly disable the shell's exit code checking for the `docker swarm leave` command (i.e. something like `set +e; docker swarm leave; set -e; ...`), but I slightly prefer the `|| true` variant for the sake of brevity.

Reply:
It seems weird to me that ansible's way of executing shells is to have `set -e`, since `set +e` is the default shell behavior normally. It seems weird to me, but I can't find documentation explicitly defining this behavior one way or another, so I'll take your word for it.

In that context, it does make sense to use the `|| true` variant. However, you don't need the `&&` in this case, and it's probably better to provide a message, as in the following multi-line variant (I think multi-line is a bit more readable):

```yaml
  shell: |
    docker swarm leave || echo "Already left swarm."
    docker swarm join --token={{ swarm_token_stdout }} {{ manager_ip }}:2377
```

Reply:
👍 Good call on the explicit message and multi-line variant. Will push up a change shortly.